File: path_planning/python2/arya/maze.py

class Deque:
    def __init__(self):
        self.__capacity = 1
        self.__deque_size = 0
        self.__first_item_index = 0
        self.__items = [None] * self.__capacity

    def __len__(self):
        return self.__deque_size

    def __str__(self):
        string = ''
        for i in range(self.__deque_size):
            string += str(self[i]) + ' '
        return string

    def __getitem__(self, index):
        return self.__items[(index + self.__first_item_index) %
                            self.__capacity]

    def __setitem__(self, index, item):
        self.__items[(index + self.__first_item_index) %
                     self.__capacity] = item

    def _resize(self):
        if self.__deque_size + 1 <= self.__capacity and \
                self.__deque_size * 3 >= self.__capacity:
            return

        new_capacity = max(1, 2 * self.__deque_size)
        new_items = [None] * new_capacity

        for i in range(self.__deque_size):
            new_items[i] = self[i]
        self.__items = new_items
        self.__capacity = new_capacity
        self.__first_item_index = 0

    def is_empty(self):
        return self.__deque_size == 0

    def add(self, index, item):
        self._resize()

        if index < self.__deque_size / 2:
            if self.__first_item_index == 0:
                self.__first_item_index = self.__capacity - 1
            else:
                self.__first_item_index -= 1

            for i in range(index - 1):
                self[i] = self[i + 1]
        else:
            for i in range(self.__deque_size, index, -1):
                self[i] = self[i - 1]
        self[index] = item
        self.__deque_size += 1

        return self

    def add_right(self, item):
        return self.add(self.__deque_size, item)

    def add_left(self, item):
        return self.add(0, item)

    def remove(self, index):
        tmp = self[index]
        if index < self.__deque_size / 2:
            for i in range(index, 0, -1):
                self[i] = self[i - 1]
            self.__first_item_index = (
                self.__first_item_index + 1) % self.__capacity
        else:
            for i in range(index, self.__deque_size - 1):
                self[i] = self[i + 1]
        self.__deque_size -= 1
        self._resize()
        return tmp

    def remove_right(self):
        return self.remove(self.__deque_size - 1)

    def remove_left(self):
        return self.remove(0)


class ArrayStack:
    def __init__(self):
        self.__stack_size = 0
        self.__capacity = 1
        self.__items = [None] * self.__stack_size

    def __len__(self):
        return self.__stack_size

    def _resize(self):
        if self.__stack_size + 1 <= self.__capacity and \
                self.__stack_size * 3 >= self.__capacity:
            return

        new_capacity = max(1, 2 * self.__stack_size)
        new_items = [None] * new_capacity

        for i in range(self.__stack_size):
            new_items[i] = self.__items[i]
        self.__items = new_items
        self.__capacity = new_capacity

    def pop(self):
        if self.__stack_size == 0:
            return None

        tmp = self.__items[self.__stack_size - 1]
        self.__stack_size -= 1
        self._resize()
        return tmp

    def peek(self):
        return self.__items[self.__stack_size - 1]

    def push(self, value):
        self._resize()
        self.__items[self.__stack_size] = value
        self.__stack_size += 1


def dfs_pathfinding(maze, initial_point, target_point):
    marked = [[False for x in maze[0]] for y in maze]
    previous_point = [[None for x in maze[0]] for y in maze]
    path_found = False
    stack = ArrayStack()
    stack.push(initial_point)

    while stack:
        current_point = stack.pop()
        if current_point == target_point:
            path_found = True
            break
        marked[current_point[0]][current_point[1]] = True

        for y in range(1, -2, -1):
            for x in range(1, -2, -1):
                neighbor_point = (current_point[0] + y, current_point[1] + x)
                if 0 <= neighbor_point[0] < len(maze) and \
                        0 <= neighbor_point[1] < len(maze[0]):
                    if maze[neighbor_point[0]][neighbor_point[1]] == 0 and \
                            not marked[neighbor_point[0]][neighbor_point[1]]:
                        stack.push(neighbor_point)
                        previous_point[neighbor_point[0]
                                       ][neighbor_point[1]] = current_point
                        marked[neighbor_point[0]][neighbor_point[1]] = True

    if not path_found:
        return None

    path = ArrayStack()
    current_point = target_point
    while current_point is not None:
        path.push(current_point)
        current_point = previous_point[current_point[0]][current_point[1]]
    return path


def main():
    height, width = raw_input().split()
    height = int(height)
    width = int(width)
    maze = []

    for row_idx in range(height):
        row = raw_input()
        row_array = []
        for block in row:
            row_array.append(int(block))
        maze.append(row_array)

    way = dfs_pathfinding(maze, (0, 0), (height - 1, width - 1))
    if way is None:
        print "That's a dead end!"
    else:
        print len(way)
        while way:
            point = way.pop()
            print str(point[0] + 1) + " " + str(point[1] + 1)


if __name__ == "__main__":
    main()

File: path_planning/cpp/arya/maze.cpp
#include <assert.h>
#include <cstring>
#include <iostream>

using namespace std;

template <class T> class Deque {
public:
  Deque(int capacity = 1);
  ~Deque();

  const T &get(int index) const;
  const T &getLast() const { return get(size - 1); }
  const T &getFirst() const { return get(0); }
  T set(int index, const T &value);

  void add(int index, const T &value);
  void addFirst(const T &value);
  void addLast(const T &value);

  T remove(int index);
  T removeFirst();
  T removeLast();

  int getSize() const { return size; }
  int getCapacity() const { return capacity; }

private:
  void resize();

  T *array;
  int size, capacity, dequeStartIndex;
};

template <class T>
Deque<T>::Deque(int capacity)
    : capacity(capacity), dequeStartIndex(0), size(0) {
  array = new T[capacity];
}

template <class T> Deque<T>::~Deque() {}

template <class T> const T &Deque<T>::get(int index) const {
  assert(index >= 0);
  assert(index < size);
  return array[(dequeStartIndex + index) % capacity];
}

template <class T> T Deque<T>::set(int index, const T &value) {
  T lastValue = get(index);
  array[(index + dequeStartIndex) % capacity] = value;
  return lastValue;
}

template <class T> void Deque<T>::add(int index, const T &value) {
  resize();

  if (index < size / 2) {
    dequeStartIndex =
        (dequeStartIndex == 0) ? capacity - 1 : dequeStartIndex - 1;
    for (int i = 0; i < index; i++) {
      array[(i + dequeStartIndex) % capacity] =
          array[(i + 1 + dequeStartIndex) % capacity];
    }
  } else {
    for (int i = size; i > index; i--) {
      array[(i + dequeStartIndex) % capacity] =
          array[(i - 1 + dequeStartIndex) % capacity];
    }
  }
  array[(index + dequeStartIndex) % capacity] = value;
  size++;
}

template <class T> void Deque<T>::addFirst(const T &value) { add(0, value); }

template <class T> void Deque<T>::addLast(const T &value) { add(size, value); }

template <class T> T Deque<T>::remove(int index) {
  assert(index < size);
  T removedValue = get(index);
  if (index < size / 2) {
    for (int i = index; i > 0; i--) {
      array[(i + dequeStartIndex) % capacity] =
          array[(i - 1 + dequeStartIndex) % capacity];
    }
    dequeStartIndex = (dequeStartIndex + 1) % capacity;
  } else {
    for (int i = index; i < size - 1; i++) {
      array[(i + dequeStartIndex) % capacity] =
          array[(i + 1 + dequeStartIndex) % capacity];
    }
  }
  size--;

  resize();

  return removedValue;
}

template <class T> T Deque<T>::removeFirst() { return remove(0); }

template <class T> T Deque<T>::removeLast() { return remove(size - 1); }

template <class T> void Deque<T>::resize() {
  if (size + 1 <= capacity && size * 3 >= capacity) {
    return;
  }

  int newCapacity = max(1, 2 * size);
  T *newArray = new T[newCapacity];
  for (int i = 0; i < size; i++) {
    newArray[i] = get(i);
  }
  // delete array;
  array = newArray;
  capacity = newCapacity;
  dequeStartIndex = 0;
}

struct Point {
  Point(int x = -1, int y = -1) : x(x), y(y) {}
  int x, y;
};

bool operator==(const Point &lhs, const Point &rhs) {
  return lhs.x == rhs.x && lhs.y == rhs.y;
}
bool operator!=(const Point &lhs, const Point &rhs) { return !(lhs == rhs); }

Deque<Point> pathfindingDFS(bool **maze, int height, int width,
                            Point initialPoint, Point targetPoint) {
  // Visited blocks
  bool **marked = new bool *[height];
  for (int y = 0; y < height; y++) {
    marked[y] = new bool[width];
    memset(marked[y], 0, sizeof(bool) * width);
  }

  Point **previous = new Point *[height];
  for (int y = 0; y < height; y++) {
    previous[y] = new Point[width];
  }

  Deque<Point> traverseQueue;
  traverseQueue.addLast(initialPoint);
  marked[initialPoint.y][initialPoint.x] = true;
  bool pathFoundFlag = false;

  while (traverseQueue.getSize() > 0) {
    Point currentPoint = traverseQueue.removeLast();
    if (currentPoint == targetPoint) {
      pathFoundFlag = true;
      break;
    }

    for (int y = 1; y >= -1; --y) {
      for (int x = 1; x >= -1; --x) {
        Point neighborPoint(currentPoint.x + x, currentPoint.y + y);
        if (neighborPoint.x >= 0 && neighborPoint.x < width &&
            neighborPoint.y >= 0 && neighborPoint.y < height) {
          if (!marked[neighborPoint.y][neighborPoint.x] &&
              maze[neighborPoint.y][neighborPoint.x]) {
            marked[neighborPoint.y][neighborPoint.x] = true;
            traverseQueue.addFirst(neighborPoint);
            previous[neighborPoint.y][neighborPoint.x] = currentPoint;
          }
        }
      }
    }
  }

  Deque<Point> path;
  if (!pathFoundFlag) {
    return path;
  }

  Point currentPoint = targetPoint;
  while (currentPoint != Point(-1, -1)) {
    path.addFirst(currentPoint);
    currentPoint = previous[currentPoint.y][currentPoint.x];
  }
  return path;
}

int main() {
  int height, width;
  cin >> height >> width;

  bool **maze = new bool *[height];
  char mazeBlock = '0';
  for (int y = 0; y < height; y++) {
    maze[y] = new bool[width];
    for (int x = 0; x < width; x++) {
      cin >> mazeBlock;
      maze[y][x] = (mazeBlock == '0') ? true : false;
    }
  }

  Deque<Point> path = pathfindingDFS(maze, height, width, Point(0, 0),
                                     Point(width - 1, height - 1));
  if (path.getSize() == 0) {
    cout << "That\'s a dead end!" << endl;
  } else {
    cout << path.getSize() << endl;
    for (int i = 0; i < path.getSize(); i++) {
      cout << path.get(i).y + 1 << " " << path.get(i).x + 1 << endl;
    }
  }

  return 0;
}
